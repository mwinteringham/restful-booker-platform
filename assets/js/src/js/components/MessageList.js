import React, { useEffect, useState } from 'react';
import { API } from '../libs/Api.js';
import Message from './Message.js';

const MessageList = ({setCount}) => {

    const [messageId, setMessageId] = useState(0);
    const [messages, setMessages] = useState([]);

    useEffect(() => {
        refreshMessageList()
    }, [])

    const deleteMessage = (id) => {
        API.deleteMessage(id, refreshMessageList);

        setCount();
    }

    const refreshMessageList = () => {
        API.getMessages(setMessages);
    }

    const openMessage = (id) => {
        setMessageId(id);
    }

    const closeMessage = () => {
        refreshMessageList();
        setMessageId(0);

        setCount();
    }

    return <div>
        {messageId > 0 && 
            <Message messageId={messageId} closeMessage={closeMessage} refreshMessageList={refreshMessageList} />
        }
        <div className="messages">
            <div className="row">
                <div className="col-sm-2 rowHeader"><p>Name</p></div>
                <div className="col-sm-9 rowHeader"><p>Subject</p></div>
                <div className="col-sm-1"></div>
            </div>
            {messages.map((value, index) => {
                return  <div className={"row detail read-" + value.read} id={"message" + index} key={index}>
                            <div className="col-sm-2" data-testid={"message" + index} onClick={() => openMessage(value.id)}><p>{value.name}</p></div>
                            <div className="col-sm-9" data-testid={"messageDescription" + index} onClick={() => openMessage(value.id)}><p>{value.subject}</p></div>
                            <div className="col-sm-1">
                            <span data-testid={"DeleteMessage" + index} className="fa fa-remove roomDelete" onClick={() => deleteMessage(value.id)}></span>
                            </div>
                        </div>
            })}
        </div>
    </div>

}

export default MessageList;
