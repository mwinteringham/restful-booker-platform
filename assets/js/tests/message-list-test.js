import React from 'react';
import MessageList from '../src/js/components/MessageList.js';
import nock from 'nock';

import '@testing-library/jest-dom'
import {
    render,
    waitFor,
    asFragment,
    fireEvent
  } from '@testing-library/react'

const messageMock = nock('http://localhost')
                        .persist()
                        .get('/message/')
                        .reply(200, {
                                messages : [
                                    {
                                        "id" : 1,
                                        "name" : "Mark Winteringham",
                                        "subject" : "Subject description here",
                                        "read" : true
                                    }, {
                                        "id" : 2,
                                        "name" : "James Dean",
                                        "subject" : "Another description here",
                                        "read" : false
                                    }, {
                                        "id" : 3,
                                        "name" : "Janet Samson",
                                        "subject" : "Lorem ipsum dolores est",
                                        "read" : true
                                    }
                                ]
                            });

test('Renders the list of messages correctly', async () => {
    const {asFragment, getByTestId} = render(<MessageList setCount={() => {}} />);

    await waitFor(() => expect(messageMock.isDone()).toBeTruthy());
    await waitFor(() => expect(getByTestId(/messageDescription0/)).toBeInTheDocument());
    await waitFor(() => expect(asFragment()).toMatchSnapshot());
});

test('Deletes message when selected to delete', async () => {
    const mock = nock('http://localhost')
                    .delete('/message/1')
                    .reply(201);

    const {getByTestId} = render(<MessageList setCount={() => {}} />);
    
    await waitFor(() => fireEvent.click(getByTestId(/DeleteMessage0/)))

    await waitFor(() => expect(mock.isDone()).toBeTruthy())
});

test('Clicking message shows message popup', async () => {
    nock('http://localhost')
        .put('/message/1/read')
        .reply(200)
    
    nock('http://localhost')
        .get('/message/1')
        .reply(200)

    const {asFragment, getByTestId} = render(<MessageList setCount={() => {}} />);

    await waitFor(() => { fireEvent.click(getByTestId(/message0/))});

    expect(asFragment()).toMatchSnapshot();
});
