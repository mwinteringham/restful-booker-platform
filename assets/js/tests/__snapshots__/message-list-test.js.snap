// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clicking message shows message popup 1`] = `
<DocumentFragment>
  <div>
    <div
      class="messages"
    >
      <div
        class="row"
      >
        <div
          class="col-sm-2 rowHeader"
        >
          <p>
            Name
          </p>
        </div>
        <div
          class="col-sm-9 rowHeader"
        >
          <p>
            Subject
          </p>
        </div>
        <div
          class="col-sm-1"
        />
      </div>
      <div
        class="row detail read-true"
        id="message0"
      >
        <div
          class="col-sm-2"
          data-testid="message0"
        >
          <p>
            Mark Winteringham
          </p>
        </div>
        <div
          class="col-sm-9"
          data-testid="messageDescription0"
        >
          <p>
            Subject description here
          </p>
        </div>
        <div
          class="col-sm-1"
        >
          <span
            class="fa fa-remove roomDelete"
            data-testid="DeleteMessage0"
          />
        </div>
      </div>
      <div
        class="row detail read-false"
        id="message1"
      >
        <div
          class="col-sm-2"
          data-testid="message1"
        >
          <p>
            James Dean
          </p>
        </div>
        <div
          class="col-sm-9"
          data-testid="messageDescription1"
        >
          <p>
            Another description here
          </p>
        </div>
        <div
          class="col-sm-1"
        >
          <span
            class="fa fa-remove roomDelete"
            data-testid="DeleteMessage1"
          />
        </div>
      </div>
      <div
        class="row detail read-true"
        id="message2"
      >
        <div
          class="col-sm-2"
          data-testid="message2"
        >
          <p>
            Janet Samson
          </p>
        </div>
        <div
          class="col-sm-9"
          data-testid="messageDescription2"
        >
          <p>
            Lorem ipsum dolores est
          </p>
        </div>
        <div
          class="col-sm-1"
        >
          <span
            class="fa fa-remove roomDelete"
            data-testid="DeleteMessage2"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Renders the list of messages correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="messages"
        >
          <div
            class="row"
          >
            <div
              class="col-sm-2 rowHeader"
            >
              <p>
                Name
              </p>
            </div>
            <div
              class="col-sm-9 rowHeader"
            >
              <p>
                Subject
              </p>
            </div>
            <div
              class="col-sm-1"
            />
          </div>
          <div
            class="row detail read-true"
            id="message0"
          >
            <div
              class="col-sm-2"
              data-testid="message0"
            >
              <p>
                Mark Winteringham
              </p>
            </div>
            <div
              class="col-sm-9"
              data-testid="messageDescription0"
            >
              <p>
                Subject description here
              </p>
            </div>
            <div
              class="col-sm-1"
            >
              <span
                class="fa fa-remove roomDelete"
                data-testid="DeleteMessage0"
              />
            </div>
          </div>
          <div
            class="row detail read-false"
            id="message1"
          >
            <div
              class="col-sm-2"
              data-testid="message1"
            >
              <p>
                James Dean
              </p>
            </div>
            <div
              class="col-sm-9"
              data-testid="messageDescription1"
            >
              <p>
                Another description here
              </p>
            </div>
            <div
              class="col-sm-1"
            >
              <span
                class="fa fa-remove roomDelete"
                data-testid="DeleteMessage1"
              />
            </div>
          </div>
          <div
            class="row detail read-true"
            id="message2"
          >
            <div
              class="col-sm-2"
              data-testid="message2"
            >
              <p>
                Janet Samson
              </p>
            </div>
            <div
              class="col-sm-9"
              data-testid="messageDescription2"
            >
              <p>
                Lorem ipsum dolores est
              </p>
            </div>
            <div
              class="col-sm-1"
            >
              <span
                class="fa fa-remove roomDelete"
                data-testid="DeleteMessage2"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="messages"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-2 rowHeader"
          >
            <p>
              Name
            </p>
          </div>
          <div
            class="col-sm-9 rowHeader"
          >
            <p>
              Subject
            </p>
          </div>
          <div
            class="col-sm-1"
          />
        </div>
        <div
          class="row detail read-true"
          id="message0"
        >
          <div
            class="col-sm-2"
            data-testid="message0"
          >
            <p>
              Mark Winteringham
            </p>
          </div>
          <div
            class="col-sm-9"
            data-testid="messageDescription0"
          >
            <p>
              Subject description here
            </p>
          </div>
          <div
            class="col-sm-1"
          >
            <span
              class="fa fa-remove roomDelete"
              data-testid="DeleteMessage0"
            />
          </div>
        </div>
        <div
          class="row detail read-false"
          id="message1"
        >
          <div
            class="col-sm-2"
            data-testid="message1"
          >
            <p>
              James Dean
            </p>
          </div>
          <div
            class="col-sm-9"
            data-testid="messageDescription1"
          >
            <p>
              Another description here
            </p>
          </div>
          <div
            class="col-sm-1"
          >
            <span
              class="fa fa-remove roomDelete"
              data-testid="DeleteMessage1"
            />
          </div>
        </div>
        <div
          class="row detail read-true"
          id="message2"
        >
          <div
            class="col-sm-2"
            data-testid="message2"
          >
            <p>
              Janet Samson
            </p>
          </div>
          <div
            class="col-sm-9"
            data-testid="messageDescription2"
          >
            <p>
              Lorem ipsum dolores est
            </p>
          </div>
          <div
            class="col-sm-1"
          >
            <span
              class="fa fa-remove roomDelete"
              data-testid="DeleteMessage2"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
